llvm/lib/IR/Instruction.cpp

```diff
@@ -177,6 +177,15 @@ void Instruction::setNonNeg(bool b) {
                          (b * PossiblyNonNegInst::NonNeg);
 }
 
+bool Instruction::hasAllowReassocOfAllOperand() const {
```

Comment:
This seems overly broad for what most checks want, which need to consider a specific operand

Reply:
Thanks , refer to https://github.com/llvm/llvm-project/blob/ebc3302725350c44aaf5f97ce7ba484e30b3efa8/llvm/include/llvm/IR/Operator.h#L310,  I think the list is as follow.
```
   case Instruction::FNeg:
    case Instruction::FAdd:
    case Instruction::FSub:
    case Instruction::FMul:
    case Instruction::FDiv:
    case Instruction::FRem:
```

Reply:
hi @arsenm ，refer to https://github.com/llvm/llvm-project/blob/ebc3302725350c44aaf5f97ce7ba484e30b3efa8/llvm/include/llvm/IR/Operator.h#L310, I'll add the operand  list as following, do you have any suggestions for that?
```
    case Instruction::FNeg:
    case Instruction::FAdd:
    case Instruction::FSub:
    case Instruction::FMul:
    case Instruction::FDiv:
    case Instruction::FRem:
```

Reply:
* I find a case in llvm\test\Transforms\InstCombine\fmul.ll, which says the reassoc flag can propagates, does it right ?
  It seems against our intentions to revise.
```
; (X*Y) * X => (X*X) * Y
; The transform only requires 'reassoc', but test other FMF in
; the commuted variants to make sure FMF propagates as expected.

define float @reassoc_common_operand1(float %x, float %y) {
; CHECK-LABEL: @reassoc_common_operand1(
; CHECK-NEXT:    [[TMP1:%.*]] = fmul reassoc float [[X:%.*]], [[X]]
; CHECK-NEXT:    [[MUL2:%.*]] = fmul reassoc float [[TMP1]], [[Y:%.*]]
; CHECK-NEXT:    ret float [[MUL2]]
;
  %mul1 = fmul float %x, %y
  %mul2 = fmul reassoc float %mul1, %x
  ret float %mul2
}
```

Reply:
hi @arsenm @jcranmer-intel @nikic 
  would you help to give some input on this?  I'm not sure if this reassoc flag can propagate this logic correctly?
 If it is correct, we should not need to check whether `the FPMathOperator of the instruction` has the reassoc flag when we try this combine.

Reply:
I think allowing the reassoc flag to spread to its neighbors is an overly aggressive interpretation. I think a strict interpretation where both instructions require the reassoc flag is the more expressive option. As in, I disagree with the premise of the comment in this test

Reply:
Thank you for your confirmation, so I'll continue to try to fix it as discussed earlier.